File: challenges/rate-limiter/php/solution/exercise.php
<?php
declare(strict_types=1);

namespace Refresher;

use Closure;
use SplQueue;

final class SlidingWindowLimiter
{
    private array $clients = [];

    public function __construct(
        private int $limit,
        private int $windowMs,
        private Closure $clock,
    ) {}

    public function allow(string $client): bool
    {
        $now = ($this->clock)();
        $timestamps = $this->clients[$client] ??= new SplQueue();
        while (!$timestamps->isEmpty() && $timestamps->bottom() <= $now - $this->windowMs) {
            $timestamps->dequeue();
        }
        if ($timestamps->count() >= $this->limit) {
            return false;
        }
        $timestamps->enqueue($now);
        return true;
    }
}

File: challenges/rate-limiter/python/solution/exercise.py
"""Reference: one accepted-timestamp deque per client."""
from collections import deque

from tooling.python.contracts import Clock


class SlidingWindowLimiter:
    def __init__(self, limit: int, window_ms: int, clock: Clock):
        self._limit = limit
        self._window_ms = window_ms
        self._clock = clock
        self._requests: dict[str, deque[int]] = {}

    def allow(self, client: str) -> bool:
        now = self._clock()
        timestamps = self._requests.setdefault(client, deque())
        while timestamps and timestamps[0] <= now - self._window_ms:
            timestamps.popleft()
        if len(timestamps) >= self._limit:
            return False
        timestamps.append(now)
        return True
